### ISSUE TITLE
Put schema in version control

### ISSUE THREAD

**Issue by MichaelCurrie** [openworm/movement_cloud#15]
We'll need to figure out how to do `mysqldump` to get a baseline version of the schema.

Subsequent changes to the schema should basically be modification statements, because the database already has data so the original code cannot simply be re-run.

**Comment by MichaelCurrie** [comment]
Other comments from @cheelee:

> Put the Schema in version control
rather than the data
It would be good too ... basically other people can choose to host their own database using only the schema.
we were discussing that the other day too.
Remote queries can be made to the base.
https://kb.iu.edu/d/aews

**Comment by MichaelCurrie** [comment]
I have added it here:

`https://github.com/openworm/movement_cloud/blob/dev/mrc_db_schema.sql`

Via the command:

`mysqldump -u root -p --no-data mrc_db4 > mrc_db_schema.sql`

But we cannot directly edit this .sql file and expect it to be reflected in the actual `mrc_db4` database.  Instead, we should incrementally change the database, and periodically re-run the `mysqldump` command to refresh the schema.  I guess?

**Comment by ver228** [comment]
I updated the database and made a pull request for the schema. I am afraid the update broke the webpage. I am sure it is because i changed the name of of some table/column. I'll try to fix it but probably it would be faster if @cheelee takes a look. If there is a bigger problem there is a backup of the old database in mrc_db4_dump_110517.sql on my home folder.

**Comment by cheelee** [comment]
Ok I can take a look at it tomorrow. Meanwhile I suggest that any future changes to the database (and the corresponding Django code) where schema or structure is concerned be done as follows:

1. Import into MySQL the new database ... call it mrc_db(x+1)
2. Point the active database in movement_site/settings.py to the new database.
3. Pull in the git dev repo.

That way if the new setup breaks, the simple way to reverse it is to roll back the git pull - it will make Django simply point back to the old database which remains intact.

**Comment by ver228** [comment]
OK. I will consider that in the future. Hopefully it is the last time the schema changes, at least for a while. Anyway, I think I fixed the problems. There were a few columns that required renaming. I made a pull request including the changes.

**Comment by MichaelCurrie** [comment]
@ver228 if you can use the remote mrc_db4 database in your local scripts, by connecting to it, then there will be no need for a local database at all.

I agree that we should have a copy we can roll back to if needed.  I suggest we have `mrc_db_dev` and `mrc_db_master`.  And we can have separate django folders pointing at the tip of the `dev` branch and `master` branches respectively, with their `settings.py` pointing to their correct respective database also.

**Comment by cheelee** [comment]
@MichaelCurrie I'm not 100% sure I'd have it set up that way. My thoughts on the dev/master separation is that they reside on different servers - master for the production server actually in real use, and dev for the developer's server, for semi-live testing prior to a push to master. 

Then for at least the dev server, we would have multiple databases - one for the latest dev pull, the others either the live database, or old ones. In my mind, I imagine that when we make such a change we make a new repo clone, pop the new database into mysql, point the new clone to the new database, and test. If successful, we simply pull dev into the live code and make it point to the new database, and delete the cloned repo. If there is a failure, nothing happens to the previously running version and we can troubleshoot using the cloned repo.

**Comment by cheelee** [comment]
@ver228 I just noticed a 122Mb sql file in your folder. Do I still use mrc_db4_dump_110517.sql to try to figure the problem out? I plan on doing this on my local machine. Makes things less messy in case more than one of us works on the dev server.

**Comment by ver228** [comment]
OK, should i delete it then? You can delete it yourself if you want.

**Comment by cheelee** [comment]
@ver228 Delete the 122Mb file? No. I was just asking you if I should be using that instead of mrc_db4_dump_110517.sql because the other file is newer. I wasn't sure hehe.

**Comment by ver228** [comment]
BTW I probably should expand the feature table with description with all the features. I will try to do it in the following days.

@MichaelCurrie I notice that now there is a mrc_db4_dev should I use this database to add new data?

**Comment by cheelee** [comment]
@ver228 mrc_db4_dev was created by me using merge_worm_db_111017_1622.sql because I thought I would be using a new database to host the new stuff. I hadn't realized you guys have both been working directly with mrc_db4 with all the changes. For now let us stick to directly changing mrc_db4 until we find a better workflow to avoid clobbering changes made by one another and stepping on each other's toes hehe.

**Comment by ver228** [comment]
OK, I agree. I  might try tonight or tomorrow.

**Comment by cheelee** [comment]
Sounds good! Meanwhile I've pushed the necessary changes to the Django tool so the database is working again. Let me know when you make changes to mrc_db4, I'll do the necessary migrations and changes when that happens. Thanks! And sorry for all the confusion earlier!

**Comment by MichaelCurrie** [comment]
I just pushed to dev: 74f2072caece881365d2a836ac1cac7f4242e128 covering the addition of the `features` table I made.

I'm glad we are sticking with manipulating `mrc_db4` directly and noting the changes via the git-controlled schema file because that workflow makes the most sense to me :)  At some point having something more complicated (like a dev and production database) might be good, if I can wrap my head around how to accomplish promotion from one dev database to a production database.

p.s. Also, if we have new issues, it is best in my opinion to create a new Issue here on GitHub rather than comment on an old issue, even if the old issue is somewhat related, because otherwise things become confusing.  :)
